### PR TITLE
[upgrade docs] Refine documentation for code upgrade

### PR DESCRIPTION
**Pull request by wrwg**
### Description

Adapt to newest implementation and extend discussion

### Test Plan

NA

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3560)

### CODE REVIEW

**Review by rajkaramchedu — APPROVED** [review]

File: developer-docs-site/docs/guides/move-guides/upgrading-move-code.md [hunk 1]
```diff
@@ -9,43 +9,111 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
...
+This restriction is intended to protect users from misuse via sharing of
+such packages. A single account can work with `arbitrary` without
+restrictions, but sharing of such code is not allowed.
 
-The upgrade policy is specified using the key `upgrade_policy`. In addition to the above-shown `upgrade_policy = "compatible"` policy, the following policies are supported:
+## Security considerations for dependencies
 
-- `immutable`: A package owner can also choose `immutable`, to disallow any future upgrade.
-- `arbitrary`: Perform no checks at all. The policy of a given package can start as `arbitrary`, then move to `compatible`, and then move to `immutable`, **but it can never move back**. This gives a consumer of a Move package verifiable guarantees on how the code is now and how it evolves in the future.
+As mentioned, even compatible upgrades can lead to harzadous effects for
```
Comment (gregnazario) [inline developer-docs-site/docs/guides/move-guides/upgrading-move-code.md:84]: Hazardous

**Review by gregnazario — APPROVED** [review]

File: developer-docs-site/docs/guides/move-guides/upgrading-move-code.md [hunk 2]
```diff
@@ -9,43 +9,111 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
...
+consumers of the upgraded code will automatically receive the new code the
+next time their code is executed. 
+
+Code upgrade enables code owners to evolve their contracts or frameworks under
+a stable, well-known account address, which can than be referenced by other
+applications, inside or outside the chain. 
+
+Code upgrade is based on an _upgrade policy_ which the owner of a package
+determines. The default policy is _(backwards) compatible_. That means, only
+those upgrades are accepted which guarantee that no existing APIs
```
[the remark this answers is not in the document]
Comment (davidiw) [reply developer-docs-site/docs/guides/move-guides/upgrading-move-code.md:22]: Might be something more nuanced such as resource layout and public api?

**Review by davidiw — APPROVED** [review]

**Comment by wrwg** [comment]
The nuances are described below this is just the summary. IMHO an API is
implicitly public?

**Comment by davidiw** [comment]
> The nuances are described below this is just the summary. IMHO an API is implicitly public?
> […](#)

And that's why you read everything around a comment...

On a more playful note, I do remember back in the day the unpublished or hidden APIs within Microsoft and other applications. And then there are other APIs that aren't accessible on a given endpoint. But yes, let's land it for tomorrow :).

File: developer-docs-site/docs/guides/move-guides/upgrading-move-code.md [hunk 3]
```diff
@@ -9,43 +9,111 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
...
 
 The Aptos chain supports _code upgrade_, which means already deployed Move
-code can be replaced with newer versions. Code upgrade enables
-code owners to evolve their contracts or frameworks under a stable, well-known
-account address, which can than be referenced by other applications, inside or
-outside the chain.
+code can be replaced with newer code. If a code upgrade happens, all 
+consumers of the upgraded code will automatically receive the new code the
+next time their code is executed. 
+
```
[the remark this answers is not in the document]
Comment (wrwg) [reply developer-docs-site/docs/guides/move-guides/upgrading-move-code.md:15]: It's actually the next transaction.

File: developer-docs-site/docs/guides/move-guides/upgrading-move-code.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (wrwg) [reply developer-docs-site/docs/guides/move-guides/upgrading-move-code.md:22]: Its actually right now both entry point functions and other public functions. Over time, we may also want to have ABI only compatibility.

**Comment by wrwg** [comment]
> > The nuances are described below this is just the summary. IMHO an API is implicitly public?
> > […](#)
> 
> And that's why you read everything around a comment...
> 
> On a more playful note, I do remember back in the day the unpublished or hidden APIs within Microsoft and other applications. And then there are other APIs that aren't accessible on a given endpoint. But yes, let's land it for tomorrow :).

Fixed.

**Pull request merged**